### PR TITLE
libxcrypt: Build only with strong hashes

### PR DESCRIPTION
**Pull request by mweinelt**
Effectively removes support for the following hashing algorithms as announced in the NixOS 22.11 release notes:

- bcrypt_x ($2x$)
- sha256crypt ($5$)
- sha1crypt ($sha1$)
- sunmd5 ($md5$)
- md5crypt ($1$)
- nt ($3$)
- bdiscrypt (_)
- bigcrypt (:)
- descrypt (:)

Passthru tests (login, shadow) built on aarch64-linux and x86_64-linux.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin (fails early during perl build)
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
Do we need to change https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/users-groups.nix#L705-L734 ?

**Comment by mweinelt** [comment]
It does not validate the scheme so far, and I don't have the energy to make it do that.

**Comment by mweinelt** [comment]
Updated the validation for `hashedPassword`. Was quite a bit simpler than expected.

cc @rnhmjoj @SuperSandro2000

File: nixos/modules/config/users-groups.nix [hunk 1]
```diff
@@ -716,7 +716,8 @@ in {
         let
           sep = "\\$";
           base64 = "[a-zA-Z0-9./]+";
-          id = "[a-z0-9-]+";
+          # must match the labels of hashes enabled in libxcrypt
+          id = "(y|gy|7|2b|2y|2a|2x|6)";
```
Comment (mweinelt) [inline nixos/modules/config/users-groups.nix:720]:
```
#name          h_prefix  nrbytes  flags
yescrypt       $y$       16       STRONG,DEFAULT,ALT,DEBIAN,FEDORA
gost_yescrypt  $gy$      16       STRONG,ALT
scrypt         $7$       16       STRONG
bcrypt         $2b$      16       STRONG,DEFAULT,ALT,FREEBSD,NETBSD,OPENBSD,OWL,SOLARIS,SUSE
bcrypt_y       $2y$      16       STRONG,ALT,OWL,SUSE
bcrypt_a       $2a$      16       STRONG,ALT,FREEBSD,NETBSD,OPENBSD,OWL,SOLARIS,SUSE
sha512crypt    $6$       15       STRONG,DEFAULT,GLIBC,FREEBSD,SOLARIS
```

https://github.com/besser82/libxcrypt/blob/v4.4.33/lib/hashes.conf#L41

File: nixos/modules/config/users-groups.nix [hunk 2]
```diff
@@ -716,7 +716,8 @@ in {
         let
           sep = "\\$";
           base64 = "[a-zA-Z0-9./]+";
-          id = "[a-z0-9-]+";
+          # must match the labels of hashes enabled in libxcrypt
+          id = "(y|gy|7|2b|2y|2a|6)";
```
Comment (mweinelt) [inline nixos/modules/config/users-groups.nix:720]:
```
#name          h_prefix  nrbytes  flags
yescrypt       $y$       16       STRONG,DEFAULT,ALT,DEBIAN,FEDORA
gost_yescrypt  $gy$      16       STRONG,ALT
scrypt         $7$       16       STRONG
bcrypt         $2b$      16       STRONG,DEFAULT,ALT,FREEBSD,NETBSD,OPENBSD,OWL,SOLARIS,SUSE
bcrypt_y       $2y$      16       STRONG,ALT,OWL,SUSE
bcrypt_a       $2a$      16       STRONG,ALT,FREEBSD,NETBSD,OPENBSD,OWL,SOLARIS,SUSE
sha512crypt    $6$       15       STRONG,DEFAULT,GLIBC,FREEBSD,SOLARIS
```

https://github.com/besser82/libxcrypt/blob/v4.4.33/lib/hashes.conf#L41

**Comment by mweinelt** [comment]
Tested with a sha256-crypt hash, and it does complain alright. Migated to yescrypt and the warning was gone.

```
trace: warning: The password hash of user "hexa" may be invalid. You must set a
valid hash or the user will be locked out of their account. Please
check the value of option `users.users."hexa".hashedPassword`.
```

**Review by rnhmjoj — COMMENTED** [review]
The changes look good but I can't properly test this right now.
Anyway, glad to see we're finally using modern hashing schemes.

File: nixos/doc/manual/release-notes/rl-2305.section.md [hunk 3]
```diff
@@ -18,6 +18,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - `nixos-rebuild` now supports an extra `--specialisation` option that can be used to change specialisation for `switch` and `test` commands.
 
+- `libxcrypt`, the library providing the `crypt(3)` password hashing function, will be built without support for algorithms not flagged [`strong`](https://github.com/besser82/libxcrypt/blob/v4.4.33/lib/hashes.conf#L48). This affects the availability of password hashing algorithms used for the system login (`login(1)`, `passwd(1)`), but also Apache2 Basic-Auth, Samba, OpenLDAP, and [many other packages](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20libxcrypt&type=code).
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2305.section.md:21]:
```suggestion
- `libxcrypt`, the library providing the `crypt(3)` password hashing function, is now built without support for algorithms not flagged [`strong`](https://github.com/besser82/libxcrypt/blob/v4.4.33/lib/hashes.conf#L48). This affects the availability of password hashing algorithms used for the system login (`login(1)`, `passwd(1)`), but also Apache2 Basic-Auth, Samba, OpenLDAP, and [many other packages](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20libxcrypt&type=code).
```

Comment (winterqt) [reply nixos/doc/manual/release-notes/rl-2305.section.md:21]: Agreed, but also:
```suggestion
- `libxcrypt`, the library providing the `crypt(3)` password hashing function, will be built without support for algorithms not flagged [`strong`](https://github.com/besser82/libxcrypt/blob/v4.4.33/lib/hashes.conf#L48). This affects the availability of password hashing algorithms used for system login (`login(1)`, `passwd(1)`), but also Apache2 Basic-Auth, Samba, OpenLDAP, and [many other packages](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20libxcrypt&type=code).
```

Comment (mweinelt) [reply nixos/doc/manual/release-notes/rl-2305.section.md]: Applied both recommendations. Thanks.

**Review by SuperSandro2000 — APPROVED** [review]

**Review by fpletz — APPROVED** [review]
Not tested yet but the diff looks fine! :+1:

**Comment by vcunat** [comment]
`libxcrypt-legacy` supporting all hashes:
https://github.com/NixOS/nixpkgs/pull/223034

**Comment by vcunat** [comment]
- [ ] `mailutils` [fails tests](https://hydra.nixos.org/build/213354268/nixlog/4/tail).  Perhaps maintainers can have a look: @orivej @vrthra
- [ ] `lighttpd` [fails tests](https://hydra.nixos.org/build/213382643/nixlog/4/tail).  Perhaps maintainers can have a look: @bjornfor @brechtcs
- [ ] `radicale2` [fails tests](https://hydra.nixos.org/build/213373432/nixlog/3/tail). Perhaps maintainers can have a look: @edwtjo @pSub @infinisil

Generally I can imagine options around updating, fixing or disabling tests and maybe even using `libxcrypt-legacy`.

**Comment by orivej** [comment]
I see that `mailutils` uses weak auth during tests to interact with its mock server, i.e. it sends plaintext `USER name`, `PASS  guessme` before sending other testing queries. Maybe we should ask its developers on the best way forward (https://savannah.gnu.org/support/?group=mailutils) and meanwhile use `libxcrypt-legacy` or disable tests.

**Comment by obadz** [comment]
I continue to think this should be configurable, see #208603

**Pull request merged**